adafruit_pm25/uart.py

```diff
@@ -46,38 +51,207 @@
 from digitalio import Direction
 from . import PM25
 
+PLANTOWER_HEADER = b"\x42\x4D"
+
+PLANTOWER_CMD_MODE_PASSIVE = b"\xE1\x00\x00"
+PLANTOWER_CMD_MODE_ACTIVE = b"\xE1\x00\x01"
+PLANTOWER_CMD_READ = b"\xE2\x00\x00"
+PLANTOWER_CMD_SLEEP = b"\xE4\x00\x00"
+PLANTOWER_CMD_WAKEUP = b"\xE4\x00\x01"
+
+UART_RETRY_COUNT = 5
+FRAME_RETRY_COUNT = 3
+MAX_FRAME_SIZE = 32
+
 
 class PM25_UART(PM25):
     """
     A driver for the PM2.5 Air quality sensor over UART
     """
 
-    def __init__(self, uart, reset_pin=None):
+    def __init__(self, uart, reset_pin=None, set_pin=None, mode="passive"):
+        self._uart = uart
+        self._reset_pin = reset_pin
+        self._set_pin = set_pin
+        self._mode = mode
+
         if reset_pin:
-            # Reset device
-            reset_pin.direction = Direction.OUTPUT
-            reset_pin.value = False
-            time.sleep(0.01)
-            reset_pin.value = True
-            # it takes at least a second to start up
-            time.sleep(1)
+            # Reset device on init, for good measure
+            self._reset_pin.direction = Direction.OUTPUT
+            self._pin_reset()
+
+        if set_pin:
+            # Pull set pin high to 'working' status (pulling low puts device to sleep)
+            self._set_pin.direction = Direction.OUTPUT
+            self._set_pin.value = True
+
+        if self._mode == "passive":
+            self._cmd_mode_passive()
+        elif self._mode == "active":
+            self._cmd_mode_active()
+        else:
+            raise RuntimeError("Invalid mode")
 
-        self._uart = uart
         super().__init__()
 
     def _read_into_buffer(self):
+        if self._mode == "passive":
+            self._cmd_passive_read()
+        self._buffer = self._read_uart()
+
+    def _cmd_mode_passive(self):
+        """
+        Sends command to device to enable 'passive' mode
+
+        In passive mode, data frames are only sent after a read command
+        """
+        self._uart.reset_input_buffer()
+        self._uart.write(self._build_cmd_frame(PLANTOWER_CMD_MODE_PASSIVE))
+        cmd_response = self._read_uart()
+        self._mode = "passive"
+        time.sleep(1)
+        return cmd_response
+
+    def _cmd_mode_active(self):
+        """
+        Sends command to device to enable 'active' mode
+
+        In active mode, data frames are sent repeatedly every second
```

Comment:
Documenting this as "every second" runs the risk of promulgating the myth that these things produce a value exactly every second which is half the reason for the bugs in libraries for this sensor.

The data sheet states:

> The active mode is divided into two sub-modes: stable mode and fast mode. If the concentration change is small the sensorwould run at stable mode with the real interval of 2.3s. And if the change is big the sensor would be changed to fast mode automatically with the interval of 200~800ms, the higher of the concentration, the shorter of the interval.

I've seen mine spitting mainly every 910ms but I'll keep an eye out for those longer and short values.

Reply:
Good catch.

I saw no large variations in the active mode reads in my testing (~900-1000ms generally) but I also didn't subject the sensors to any large variations during testing either. So this quirk in the docs became quickly forgotten. :joy:

This should definitely be changed in the comments to provide more clarity or point to documentation that can.

Reply:
Addressed this in f66319a. Tried to capture the doc's general gist in a short explanation.

Let me know if that looks better.

Reply:
Looks good.